File: api/app/Services/LedgerService.php
<?php
// app/Services/LedgerService.php

namespace App\Services;

use App\Models\AccountTransaction;
use App\Models\Cashbox;
use App\Models\CashboxTransaction;
use App\Models\CurrentAccount;
use Illuminate\Support\Facades\DB;

/**
 * Para hareketleri tek DB transaction'ında atomik yazılır (v1.1 26.5).
 * Bakiye kolonları cache'tir; gerçeklik hareket tablolarındadır.
 */
class LedgerService
{
    /**
     * Tahsilat: müşteri borcunu azaltır (cari credit) + kasaya giriş (cashbox in).
     */
    public function collect(int $customerId, float $amount, string $paymentMethod, int $cashboxId, ?string $description, ?int $userId): array
    {
        return DB::transaction(function () use ($customerId, $amount, $paymentMethod, $cashboxId, $description, $userId) {
            $account = CurrentAccount::firstOrCreate(['customer_id' => $customerId], ['balance' => 0]);
            $cashbox = Cashbox::lockForUpdate()->findOrFail($cashboxId);
            $account = CurrentAccount::lockForUpdate()->find($account->id);

            // Cari: borç azalır
            $account->balance = (float) $account->balance - $amount;
            $account->save();

            $accTx = AccountTransaction::create([
                'account_id'     => $account->id,
                'type'           => 'credit',
                'amount'         => $amount,
                'balance_after'  => $account->balance,
                'description'    => $description ?? 'Tahsilat',
                'source_type'    => 'collection',
                'cashbox_id'     => $cashboxId,
                'payment_method' => $paymentMethod,
                'created_by'     => $userId,
                'created_at'     => now(),
            ]);

            // Kasa: giriş
            $cashbox->balance = (float) $cashbox->balance + $amount;
            $cashbox->save();

            CashboxTransaction::create([
                'cashbox_id'    => $cashbox->id,
                'type'          => 'in',
                'amount'        => $amount,
                'balance_after' => $cashbox->balance,
                'description'   => $description ?? 'Tahsilat',
                'source_type'   => 'collection',
                'source_id'     => $accTx->id,
                'created_by'    => $userId,
                'created_at'    => now(),
            ]);

            return [
                'account_balance' => $account->balance,
                'cashbox_balance' => $cashbox->balance,
                'transaction_id'  => $accTx->id,
            ];
        });
    }

    /** Kasalar arası transfer (atomik). */
    public function transfer(int $fromId, int $toId, float $amount, ?int $userId): array
    {
        abort_if($fromId === $toId, 422, 'Aynı kasaya transfer yapılamaz.');

        return DB::transaction(function () use ($fromId, $toId, $amount, $userId) {
            // Kilit sırasını id'ye göre sabitle (deadlock önleme)
            [$firstId, $secondId] = $fromId < $toId ? [$fromId, $toId] : [$toId, $fromId];
            Cashbox::lockForUpdate()->whereIn('id', [$firstId, $secondId])->get();

            $from = Cashbox::findOrFail($fromId);
            $to   = Cashbox::findOrFail($toId);

            abort_if((float) $from->balance < $amount, 422, 'Kaynak kasada yeterli bakiye yok.');

            $from->balance = (float) $from->balance - $amount;
            $from->save();
            $to->balance = (float) $to->balance + $amount;
            $to->save();

            CashboxTransaction::create([
                'cashbox_id' => $from->id, 'type' => 'transfer', 'amount' => $amount,
                'balance_after' => $from->balance, 'description' => "Transfer → {$to->name}",
                'source_type' => 'transfer', 'transfer_to_id' => $to->id,
                'created_by' => $userId, 'created_at' => now(),
            ]);
            CashboxTransaction::create([
                'cashbox_id' => $to->id, 'type' => 'in', 'amount' => $amount,
                'balance_after' => $to->balance, 'description' => "Transfer ← {$from->name}",
                'source_type' => 'transfer', 'transfer_to_id' => $from->id,
                'created_by' => $userId, 'created_at' => now(),
            ]);

            return ['from_balance' => $from->balance, 'to_balance' => $to->balance];
        });
    }
}

File: api-net/Services/LedgerService.cs
using Liftonom.Api.Data;
using Liftonom.Api.Models;
using Microsoft.EntityFrameworkCore;

namespace Liftonom.Api.Services;

/// <summary>Para hareketleri tek DB transaction'ında atomik yazılır (Laravel LedgerService karşılığı).</summary>
public class LedgerService(AppDbContext db)
{
    public record CollectResult(decimal AccountBalance, decimal CashboxBalance, long TransactionId);

    /// <summary>İşi bir transaction içinde çalıştırır; çağıran zaten bir transaction açtıysa onu kullanır.</summary>
    private async Task<T> RunAsync<T>(Func<Task<T>> work)
    {
        if (db.Database.CurrentTransaction != null) return await work();
        await using var tx = await db.Database.BeginTransactionAsync();
        var result = await work();
        await tx.CommitAsync();
        return result;
    }

    /// <summary>Tahsilat: cari borç azalır + kasaya giriş.</summary>
    public Task<CollectResult> CollectAsync(long customerId, decimal amount, string paymentMethod,
        long cashboxId, string? description, long? userId) => RunAsync(async () =>
    {
        var account = await db.CurrentAccounts.FirstOrDefaultAsync(a => a.CustomerId == customerId);
        if (account == null)
        {
            account = new CurrentAccount { TenantId = db.CurrentTenantId!.Value, CustomerId = customerId, Balance = 0,
                CreatedAt = DateTime.UtcNow, UpdatedAt = DateTime.UtcNow };
            db.CurrentAccounts.Add(account);
            await db.SaveChangesAsync();
        }

        var cashbox = await db.Cashboxes.FirstOrDefaultAsync(c => c.Id == cashboxId)
            ?? throw new ApiException(404, "Kasa bulunamadı.");

        account.Balance -= amount;
        account.UpdatedAt = DateTime.UtcNow;

        var accTx = new AccountTransaction
        {
            TenantId = db.CurrentTenantId!.Value, AccountId = account.Id, Type = "credit", Amount = amount,
            BalanceAfter = account.Balance, Description = description ?? "Tahsilat", SourceType = "collection",
            CashboxId = cashboxId, PaymentMethod = paymentMethod, CreatedBy = userId, CreatedAt = DateTime.UtcNow,
        };
        db.AccountTransactions.Add(accTx);

        cashbox.Balance += amount;
        cashbox.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync();

        db.CashboxTransactions.Add(new CashboxTransaction
        {
            TenantId = db.CurrentTenantId!.Value, CashboxId = cashbox.Id, Type = "in", Amount = amount,
            BalanceAfter = cashbox.Balance, Description = description ?? "Tahsilat", SourceType = "collection",
            SourceId = accTx.Id, CreatedBy = userId, CreatedAt = DateTime.UtcNow,
        });
        await db.SaveChangesAsync();

        return new CollectResult(account.Balance, cashbox.Balance, accTx.Id);
    });

    /// <summary>Kasalar arası transfer.</summary>
    public Task<(decimal FromBalance, decimal ToBalance)> TransferAsync(long fromId, long toId, decimal amount, long? userId)
    {
        if (fromId == toId) throw new ApiException(422, "Aynı kasaya transfer yapılamaz.");
        return RunAsync(async () =>
        {
            var from = await db.Cashboxes.FirstOrDefaultAsync(c => c.Id == fromId) ?? throw new ApiException(404, "Kaynak kasa bulunamadı.");
            var to = await db.Cashboxes.FirstOrDefaultAsync(c => c.Id == toId) ?? throw new ApiException(404, "Hedef kasa bulunamadı.");
            if (from.Balance < amount) throw new ApiException(422, "Kaynak kasada yeterli bakiye yok.");

            from.Balance -= amount; from.UpdatedAt = DateTime.UtcNow;
            to.Balance += amount; to.UpdatedAt = DateTime.UtcNow;
            var now = DateTime.UtcNow;

            db.CashboxTransactions.Add(new CashboxTransaction { TenantId = db.CurrentTenantId!.Value, CashboxId = from.Id,
                Type = "transfer", Amount = amount, BalanceAfter = from.Balance, Description = $"Transfer → {to.Name}",
                SourceType = "transfer", TransferToId = to.Id, CreatedBy = userId, CreatedAt = now });
            db.CashboxTransactions.Add(new CashboxTransaction { TenantId = db.CurrentTenantId!.Value, CashboxId = to.Id,
                Type = "in", Amount = amount, BalanceAfter = to.Balance, Description = $"Transfer ← {from.Name}",
                SourceType = "transfer", TransferToId = from.Id, CreatedBy = userId, CreatedAt = now });

            await db.SaveChangesAsync();
            return (from.Balance, to.Balance);
        });
    }

    /// <summary>Kasadan çıkış (gider/maaş).</summary>
    public Task<decimal> CashOutAsync(long cashboxId, decimal amount, string? description, string sourceType, long? sourceId, long? userId)
        => RunAsync(async () =>
    {
        var cashbox = await db.Cashboxes.FirstOrDefaultAsync(c => c.Id == cashboxId) ?? throw new ApiException(404, "Kasa bulunamadı.");
        if (cashbox.Balance < amount) throw new ApiException(422, "Kasada yeterli bakiye yok.");

        cashbox.Balance -= amount; cashbox.UpdatedAt = DateTime.UtcNow;
        db.CashboxTransactions.Add(new CashboxTransaction { TenantId = db.CurrentTenantId!.Value, CashboxId = cashbox.Id,
            Type = "out", Amount = amount, BalanceAfter = cashbox.Balance, Description = description,
            SourceType = sourceType, SourceId = sourceId, CreatedBy = userId, CreatedAt = DateTime.UtcNow });

        await db.SaveChangesAsync();
        return cashbox.Balance;
    });

    /// <summary>Tahsilat iptali/iadesi: cari +tutar (ters debit), kasa −tutar (out). Çift iade engellenir.</summary>
    public Task<CollectResult> ReverseCollectionAsync(long accountTxId, long? userId) => RunAsync(async () =>
    {
        var orig = await db.AccountTransactions.FirstOrDefaultAsync(t => t.Id == accountTxId)
            ?? throw new ApiException(404, "Tahsilat hareketi bulunamadı.");
        if (orig.SourceType != "collection") throw new ApiException(422, "Bu hareket bir tahsilat değil.");
        if (await db.AccountTransactions.AnyAsync(t => t.SourceType == "collection_reversal" && t.SourceId == orig.Id))
            throw new ApiException(422, "Bu tahsilat zaten iade edilmiş.");

        var account = await db.CurrentAccounts.FirstOrDefaultAsync(a => a.Id == orig.AccountId)
            ?? throw new ApiException(404, "Cari hesap bulunamadı.");
        var now = DateTime.UtcNow;
        var tid = db.CurrentTenantId!.Value;

        account.Balance += orig.Amount; account.UpdatedAt = now;
        db.AccountTransactions.Add(new AccountTransaction
        {
            TenantId = tid, AccountId = account.Id, Type = "debit", Amount = orig.Amount, BalanceAfter = account.Balance,
            Description = $"Tahsilat iadesi (#{orig.Id})", SourceType = "collection_reversal", SourceId = orig.Id,
            CashboxId = orig.CashboxId, PaymentMethod = orig.PaymentMethod, CreatedBy = userId, CreatedAt = now,
        });

        decimal cashboxBalance = 0;
        if (orig.CashboxId is { } cbId)
        {
            var cb = await db.Cashboxes.FirstOrDefaultAsync(c => c.Id == cbId);
            if (cb != null)
            {
                cb.Balance -= orig.Amount; cb.UpdatedAt = now;
                db.CashboxTransactions.Add(new CashboxTransaction
                {
                    TenantId = tid, CashboxId = cb.Id, Type = "out", Amount = orig.Amount, BalanceAfter = cb.Balance,
                    Description = $"Tahsilat iadesi (#{orig.Id})", SourceType = "collection_reversal", SourceId = orig.Id,
                    CreatedBy = userId, CreatedAt = now,
                });
                cashboxBalance = cb.Balance;
            }
        }
        await db.SaveChangesAsync();
        return new CollectResult(account.Balance, cashboxBalance, orig.Id);
    });
}
